lib/rack/lint.rb

```diff
@@ -179,6 +180,11 @@ def check_environment(env)
         ##                        to +call+ that is used to perform a full
         ##                        hijack.
 
+        ## <tt>rack.protocol</tt>:: If the request is an HTTP/1 upgrade or
```

Comment:
I'm not in favor of the `rack.protocol` naming.  It seems like it would be which HTTP version is in use (`SERVER_PROTOCOL`) or `http` or `https` (`rack.url_scheme`).

I assume the HTTP/1 upgrade header is already available in `HTTP_UPGRADE`.  Is the HTTP/2 protocol pseudo header exposed anywhere yet?

Absent a better idea, I recommend switching to `rack.upgrade_protocol`, since it looks this this is used for upgrading the protocol used for the connection.

Regardless of what we select, I think we want to explicitly specify that this in an array (even if it can only contain one value in certain HTTP versions), since it looks like we specify that the env key should be an array later.

Reply:
HTTP/2 literally called this the `:protocol` pseudo header. I actually don't have a strong opinion about the name but I've already used this name for my current implementation of websocket simply because I had to choose something and it seemed like the logical approach.

I also checked https://datatracker.ietf.org/doc/html/rfc3875 and could find no reference about how to handle `connection: upgrade` requests. It doesn't seem like the CGI specification which we follow for our `env` keys has any guidance on the matter.

`rack.upgrade_protocol` - I would suggest we avoid this, or the "upgrade" terminology in general, as HTTP/2 explicitly calls this out in a number of places as being unsupported and I think the word "upgrade" is deprecated in favour of "connect" and "protocol".

- https://www.rfc-editor.org/rfc/rfc9113.html#name-the-upgrade-header-field
- https://www.rfc-editor.org/rfc/rfc8441.html#section-1

The specific introduction of the protocol pseudo header is here: https://www.rfc-editor.org/rfc/rfc8441.html#section-4

I'm fine with specifying that it should be an `Array`.

Reply:
`rack.connect_protocol` maybe?  `rack.protocol` is too confusing as a name for this, IMO.

I would still like to know how changing the connection protocol is currently handled for HTTP/2 in rack.  Is it impossible with the current design, does it require server-specific code, or is it possible and just implemented a different way?

Reply:
`rack.protocol` is the intersection of all the different semantics. `connect_protocol` is too specific to HTTP/2 that it makes the HTTP/1 use case clumsy. It doesn't make a lot of sense to have `rack.connect_protocol` in the response headers map to an HTTP/1 connection upgrade IMHO. I understand the value of being more specific, but this isn't all or nothing, the intersection between HTTP/1 and HTTP/2 semantics is very wide and doesn't invite more specific naming IMHO. In other words, being too specific in the naming is equally confusing.

Reply:
I probably should have been more specific. There is nothing that ties the `:protocol` pseudo header to the `CONNECT` method, so linking them as `rack.connect_protocol` would be a bit clumsy if later on another method was introduced that also used the `:protocol` pseudo header with a different method.